### DISCUSSION TITLE
Please reconsider the decision to (temporarily?) abandon the Remix name and instead release Remix v3 as a thin wrapper around React Router 7

### DISCUSSION THREAD

**Discussion by torbensky** [remix-run/remix#10333] [General]
I get that all the functionality you want is now supported in react router v7, but I don't understand why this has to mean "abandoning" the Remix name and orphaning the ecosystem that has evolved around it? 

It feels to me like this is going to generate a lot of unnecessary confusion that could be avoided by continuing to support the **Remix** name.

As someone who just recently adopted Remix v2, I feel so confused by this decision.

What's even more confusing is that the plan is to later return to the Remix name at some future date?

> As for the Remix brand, it's not going anywhere. We are the Remix team, React Router is a Remix project, and we have really exciting plans beyond React Router we can't wait to talk about. The Remix packages are going to take a little nap. Right now the Remix team is going to be heads down shipping React Router v7 and delivering the smoothest upgrade process we possibly can.

So you're asking Remix adopters to migrate `Remix -> React Router -> Remix` at some future date? Or are you going to turn "Remix" into something else? It's a weird identify crisis.

As mentioned [in the blog post](https://remix.run/blog/merging-remix-and-react-router)

> Actually, we're doing a little more than that. Remix has always just been a layer on top of React Router - and that layer has been shrinking over time. It's now so small that we're just going to eliminate it.

Most of Remix was already a thin wrapper over React Router. So why not simply continue wrapping React Router with Remix for v3? If you're not actually ending the Remix project, wouldn't this make it less confusing when you eventually release the next iteration of Remix that is once again different from React Router?

**TLDR;** I feel like this decision isn't appreciating the value of the name "Remix" and the ecosystem and SEO that has evolved around that name and I as as a "Remix" user feel abandoned.

**Comment by gustavopch** [comment]
It wouldn't make sense to keep 2 names for the same thing.

People writing blog posts, making YouTube videos, etc. would always have to refer to both names for SEO. On the other hand, people searching for articles and videos would also always have to search for both names. And then you have 2 documentations, 2 websites, 2 repos with duplicated issues, etc.

Only difference between Remix and React Router was the Vite plugin. React Router was a library, Remix was a framework, but React 19 can't be fully used if you don't have a framework to implement the RSC stuff. What should they do? Keep React Router incompatible with RSC? Convince React Router users to migrate to Remix and abandon the brand that took a whole decade for them to build? Let's be honest, there simply wasn't any path forward without its trade-offs. They took the pragmatic choice. The only downside from this choice is that for some time people will be confused (TBH, I don't understand why people find it so confusing), but it's a temporary problem affecting only Remix users (which is only a fraction compared to all React Router users) and that will fade away with time — while all other choices would be permanent source of problems.

> So you're asking Remix adopters to migrate Remix -> React Router -> Remix at some future date?

That's certainly not the idea. There's rumor they'll reuse the Remix name in the future for a completely new thing. I'm particularly hoping it's a Rails-like backend framework, but no idea.

**Reply by torbensky** [reply]
> There's rumor they'll reuse the Remix name in the future for a completely new thing.

It's not rumor, that came from their official blog post.

**Reply by torbensky** [reply]
> We can't talk about it much (yet!), but we have ideas to make Remix something more powerful and even more server-centric, something you'd use a React Router project on top of.

source: [Merging Remix and React Router](https://remix.run/blog/merging-remix-and-react-router)

It just feels like a neither here nor there decision. Like the idea is to both end the project but not end it.

> TBH, I don't understand why people find it so confusing

I think it's very confusing to hear that the project is both being discontinued but not at the same time.

For what it's worth, I do get that no decision is going to make everyone happy. So take this for what it is: the genuine feedback of a user who is coming with the perspective of being a recent Remix adopter. I have been super excited about **Remix** and bought into that project and idea. This decision makes me doubt the direction and stability of the ecosystem and is making me seriously reconsider investing my skills in another framework.

**Reply by torbensky** [reply]
>  Remix is coming back better than ever in a future release with an incremental adoption strategy enabled by these changes

[Incremental Path to React 19: React Conf Follow-Up](https://remix.run/blog/incremental-path-to-react-19)

Another reference to the fact that the official plan is to continue the Remix project, coming from the official blog, not a rumor.

**Reply by gustavopch** [reply]
> It's not rumor, that came from their official blog post.

I mean, there's no confirmation that it will actually happen. They could decide to use a different name for the new thing.

> I think it's very confusing to hear that the project is both being discontinued but not at the same time.

The thing you knew as Remix is now called React Router. As simple as that. The software is not discontinued, it just has a different name now.

> This decision makes me doubt the direction and stability of the ecosystem and is making me seriously reconsider investing my skills in another framework.

Sorry, but this also doesn't make sense. React Router has been out there for more than a decade. They could have decided to abandon it and keep only Remix. THAT would be a reason to question the long-term stability of the software made by this team. Their decision to keep maintaining React Router is the opposite of a reason to raise that question. Especially as neither Remix nor React Router users are being left behind. It's literally a matter of naming. It's still the same piece of software. Just like how you can call the language ECMAScript and JavaScript. Someday, hopefully, we'll be able to stop using the name ECMAScript and call it just JavaScript.

**Reply by torbensky** [reply]
> Just like how you can call the language ECMAScript and JavaScript.

Yes and that is also really confusing. The ECMAScript name was an unfortunate result of trademark issues. Not at all like the Remix decision.

Anyway, I understand what you're saying, but I disagree with your conclusion. I think the name matters. I think that telling people to migrate off Remix while also keeping that name around to possibly become something entirely different is confusing.

I think at the very least, if the plan is to move people off Remix to react-router, then Remix should be ended. I think that the decision to sort of park the name to maybe reuse it for something entirely different later is even more unnecessary and not good for us users.

**Reply by gustavopch** [reply]
> Not at all like the Remix decision.

I mean, both are cases of a single thing having 2 names; of course, for very different reasons.

> I think the name matters.

That's why the React Router name was the pragmatic choice. It's much more well known name. It would take them another decade to have any chance of making the Remix name as well known as the React Router name *already is*.

> then Remix should be ended

Particularly, I don't care if they release the new thing as "Remix v3", but I wouldn't be surprised if they decided to name it something else (Reverb? [literally-everything](https://x.com/ryanflorence/status/1864086550113579246)? lol).

**Reply by torbensky** [reply]
Another way to look at it is that Remix had its own, clear identity that was free of React Router's history and this allowed the Remix project to forge its own path and focus on the needs of Remix users. Now it's being swallowed by a much larger community that may have different needs and priorities. And it's causing some confusion in the process. 

But I can see how it's easier for the maintainers to do this and why they would want to prioritize a larger community over a smaller one.

**Reply by gustavopch** [reply]
Indeed, they wouldn't have the freedom to experiment all the new ideas if they didn't start a separate project. But many of the ideas were adopted by React itself, so there isn't a lot left to experiment regarding data and routing. Merging back into React Router means we still have what we like in Remix, but now in a much larger ecosystem, with many more people filling issues, contributing to its stability, building libraries around it, etc.

**Reply by aarjithn** [reply]
Well, its certainly possible to call it different names - Tanstack is doing this with router (library) and start (framework)

They even explain it clearly here https://tanstack.com/router/latest/docs/framework/react/start/overview#router-or-start

When RR team says "no breaking changes" it is not the case as users have to rename imports to nee @react-router scope and community have to rename packages..

**Comment by torbensky** [comment]
fwiw I think Remix is an amazing project and y'all did fantastic work and I'm thankful for that :)

**Comment by licongy** [comment]
I'm very confused now, should I use that codemod to give up remix completely right now? I don't think any project team would continue to use a framework that has an uncertain future, is being abandoned, and doesn't know when it will be picked up again. Unless the team has a clear roadmap for the future of Remix, which it doesn't seem to have at the moment, then it should be reasonable to assume there won't be a v3 anymore.

And the fact that codemod can't solve anything except help us some simple replace job, with make lots errors. Going from remix to RR7 seems to require a lot of refactoring, but I don't even know which site to open if I want to look up references/docs. What a huge disaster.

**Reply by gustavopch** [reply]
> I don't think any project team would continue to use a framework that has an uncertain future, is being abandoned, and doesn't know when it will be picked up again.

People are not supposed to keep using Remix. Quite the opposite. The idea is exactly that, once you have all future flags enabled, change your imports to RR7. It's really that simple. There's literally nothing else besides this. The software is the same, just with a different name. It was called Remix, now it's called React Router.

**Reply by licongy** [reply]
I think we can all pretty much agree that both RR7 and Remix are excellent. And if they're already so similar, a merge makes sense. But! The problem is the suddenness, haste, and lack of transparency. The team gave us high expectations for Remix v3 from the middle of the year, and now this sudden betrayal is hurting the supporters of this hard-to-establish brand in some way, and it's not wise to deliberately ignore and underestimate this impact.

**Reply by gustavopch** [reply]
But it's really just a name change. Only the name "Remix v3" was abandoned. The thing that that name references wasn't.

**Reply by miran248** [reply]
It's not just the name / import change, configs and clis have changed as well!
Also, before the change, you had react-router and remix, now you have react-router (library mode) and react-router (framework mode).
Before you had remix cli and later vite, now you have react-router, which is just a thin wrapper around vite.
This new setup also requires [yet another config file](https://github.com/remix-run/react-router/blob/main/decisions/0013-react-router-config-ts.md) (should have just stayed as a vite plugin, we were so close).
It's exhausting, I started three remix projects over the past 12 months and the setup was totally different every time!

That said, I still like it and will continue using it, won't recommend it though, at least not until the dust settles.

**Reply by gustavopch** [reply]
> It's not just the name / import change, configs and clis have changed as well!

Changing the name of the project implies renaming things like the CLI from `remix` to `react-router`. It's still part of the name change, not *something else*.

> This new setup also requires yet another config file

I'd rather not have a new file too, but this has no relation to the name change. It would just be called `remix.config.ts` if they didn't rename the project.

> It's exhausting, I started three remix projects over the past 12 months and the setup was totally different every time!

True. OTOH, it's hard to improve something if you can't change what you need to change to improve it. 🤷‍♂️

**Comment by HriBB** [comment]
I migrated 5 od my smaller (10 routes) remix v2 apps to react router 7 in a matter of minutes. First migration took maybe 40min, next one was 15min. 

Will try and update another bigger app (40 routes) shortly, but I am pretty confident, it wont take more than 2 or 3h, because I used programatic routes instead of file based routing.

I love what the Remix/React Router team is doing.

**Comment by m1rn** [comment]
Yeah, I’m a bit confused about this too. React Router has always been about providing solid routing solutions, but now it seems to be moving more towards a framework-style approach. Based on the name, I feel like React Router should stick to its core purpose—routing—and leave the framework stuff to Remix. What do you guys think about this direction?

**Reply by gustavopch** [reply]
If I understand correctly, that's not a possible direction in the paradigm of React v19 as only a framework can implement RSC.

**Comment by niels-van-den-broeck** [comment]
Weighing in a bit late but I feel like this might really hurt discoverability and developer experience. 
I've had a bit of a rocky relationship with React Router over the years. Googling how to do something gives wildly different results every time due to the numerous API changes over the years. This will only make things even harder to find.

But there wasn't really any competition doing things better so 🤷 

Note: This is by no means a jab at the decisions which have been taken. Just feedback

**Reply by gustavopch** [reply]
This was caused by Remix and React Router being mostly similar things. In many occasions, the documentation of Remix would just point to the documentation of React Router because some features were just re-exports. Unifying them is actually the solution.

**Reply by niels-van-den-broeck** [reply]
I understand your point of view. But the docs are by no means the only source of help. I think you had a tough decision to make and you made it.

I don't think I agree with the decision but that's fine. I don't have the years of experience working on the repo that you guys do ;)

**Reply by gustavopch** [reply]
> But the docs are by no means the only source of help.

The same problem would happen with articles, etc. People using Remix would write articles that should be helpful to React Router users and vice-versa, but they'd not reach the 2 audiences because of the two different names. As a writer, you'd frequently have to include both "Remix" and "React Router" for SEO; as a reader, you'd frequently have to do each search with the 2 terms as well. Unifying into a single name will heal this, but it will take some time.

Also worth mentioning React Router would probably not be able to adopt the new RSC paradigm introduced by React v19 without a compiler — and the lack of a compiler was essentially the only difference between React Router and Remix.

PS: I'm not part of the team, just echoing what I can understand from their motivations.

**Comment by estenp** [comment]
First off, let's be honest. The path forward is set, so this is probably a moot discussion. But that said, I tend to agree with the frustrations and confusion around it. I can't help but feel like the wools being pulled over a bit...

They needed to give CRA + RR users a bridge to modern React (or abandon that massive user base). I guess RR is too prevalent to deprecate in favor of Remix, so they went the other direction. 

As a Remix user, _Remix_ is the modern React tool, so that is a little frustrating, but I get it. The React ecosystem changed since Remix became a thing, and that has implications on the future of RR, and it's users.

But is mitigating the impact on RR users vs Remix users the full story? What makes me feel uneasy are the vague hints of what is to come for the "Remix" name. The team wants to release a new meta/fullstack framework for the modern era, continuing this ethos of the "Remix" product filling that role.

In other words, as a Remix user, who adopted with the idea that I was on this team's flagship product, how could I not feel like this tool is being deprecated if it's merging into the legacy tool, and there is plans to release a new product under "Remix"?

We don't know what they actually have planned for "Remix" and thats the exact problem. Maybe it's something I'd never want to use... or maybe it is going to be what I thought I'm building my app rewrite on now - the teams flagship full stack framework. I guess I just have to keep my head down and continue building my rewrite on a decade old Frankenstein library and hope we shouldn't have used NextJS way back when.

**Reply by gustavopch** [reply]
I think it would be cool if the new Remix were a Rails-like framework. It would still be something that complements React Router, including more cool stuff built around it. OTOH, if the new Remix is just a bunch of utility libraries, it will feel weird to me too.

**Comment by seb-lean** [comment]
Fwiw I'm currently picking a tech stack for greenfield dashboard project at our company, Remix would be my top choice, but now it's going to the bottom of the list. 

Reason: It's very difficult to explain to my team and managers what it is. The Remix site is very convincing and understandable for it's benefits over competitors, but now I have to tell stakeholders we'd actually be using Router 7, which succeeds Remix, but has a different name, but is actually the same thing, oh and yeah the website isn't as good, just look the Remix one. It's a mess. 

Obviously people will look at Next.js and say "I want that one".

**Reply by miran248** [reply]
They should have just kept things as they were. There's a huge overlap between remix and react router but they serve different purposes. If anything it should have been the other way around, considering react 19 now wants to be a _full-stack_ framework.

**Reply by gustavopch** [reply]
> There's a huge overlap between remix and react router but they serve different purposes.

Basically, the only difference between React Router v6 and Remix v2 was the Vite plugin, which is what made RR a library and Remix a framework. To support Server Components, RR would also need to become a framework anyway, so that difference wouldn't possibly exist anymore.

> [...] considering react 19 now wants to be a full-stack framework.

React v19 doesn't want to be used on its own as a framework. It wants to be used *with a framework*, which is the reason they didn't want to recommend Vite because "you'd have to build your own framework".

---

Particularly, I can only see 4 possible paths they could have taken:
1. Keep RR as a library, incompatible with RSC, and require all RR users to migrate to Remix if they want to use RSC.
2. Make RR a framework and keep both RR and Remix separate, one of which would essentially be an index file re-exporting the other.
3. Merge them and call it RR.
4. Merge them and call it Remix.

**Reply by licongy** [reply]
> 
> 1. Keep RR as a library, incompatible with RSC, and require all RR users to migrate to Remix if they want to use RSC.
> 2. Make RR a framework and keep both RR and Remix separate, one of which would essentially be an index file re-exporting the other.
> 3. Merge them and call it RR.
> 4. Merge them and call it Remix.

If we let the community vote, I'm guessing #4 would win because Remix really is so much better sounding, writing and recognisable than RR. 

I now have to even describe RR to any AI model in Cursor with "...RR7(alternative name for Remix)" in order to get correct/better answer. It's probably because Remix's docs is much better than RR's. But unfortunately, the Remix docs doesn't exactly apply to RR7 either. It's a mess.

**Reply by miran248** [reply]
> React v19 doesn't want to be used on its own as a framework. It wants to be used with a framework

Correct, apologies for the poor choice of words.

Since remix was already a framework, fourth option would make more sense, but not everyone needs RSC, thus first option would be the least disruptive..

**Reply by gustavopch** [reply]
The first and fourth options would mean abandoning a brand that took a decade to build. The Remix brand would probably never become as popular as RR *already is* given there are many competitors these days.

With time, the differences between RR and Remix would accumulate and more RR users would start migrating to other solutions instead of Remix when they want to try RSC. RR would perish and Remix would probably never fill its position.

The choice they made means RR, a decade-old piece of software, is now ready to be useful for another decade. It also means that the software we used to call Remix will have a lot more users. With more users, it will have more people battle-testing it, identifying edge cases, and submitting issues and PRs.

**Reply by licongy** [reply]
@gustavopch I'm glad you joined the discussion to give us an indirect idea of what the dev team thinks. But I don't understand why dev team is absent in this thread, as hiding behind you to avoid coming to this question that is difficult to answer by themself

**Reply by gustavopch** [reply]
Yeah, good point, I'm giving my opinion as a user, so don't take it as any kind of semi-official answer. I don't think they're exactly "hiding behind me" because TBH, I don't even think they're aware of this discussion. 😅 I'd also like them to show up and give an official answer to either corroborate or correct my points — and also explain what's planned for Remix v3, which we know will exist as a "new thing".

**Reply by aarjithn** [reply]
@gustavopch Could have also kept ‘react-router’ for everything in browser runtime and instead of inventing @react-router scope use them under @remix-run, like @remix-run/vite-plugin. 

being a npm scope means it could also host things like multipart parser or reverb alongside vite plugin.

this is exactly what tanstack router/start is doing.

**Reply by quantuminformation** [reply]
@anon-ghub vanillajs will remain stable ;) https://github.com/quantuminformation/vanillajs-patterns

**Comment by anhdd-kuro** [comment]
Remix is owned by Shopify now. I don't think they will easily give up that brand.
And you got your answer now
[remix-declaration-of-independence](https://github.com/kayac-chang/remix-website/blob/44b40299fd364b4e825e66f747d3f6cf3dddd884/data/posts/remix-declaration-of-independence.md)

For who too lazy to read, here is the main points:

> We're simply struggling to build something that feels like React did when we first found it.
> That's why Remix is moving on from React
> Remix v3 is a completely new thing. It's our fresh take on simplified web development with its own rendering abstraction in place of React
> The Remix team will continue to maintain and improve React Router
> We don't intend for React Router apps to migrate to Remix v3, they are different things
> Additionally, Remix is modular, so a lot of it will be usable from within React Router

**Comment by MichaelDeBoey** [comment]
There's now an official blogpost about the future of Remix: https://remix.run/blog/wake-up-remix

**Reply by miran248** [reply]
Sounds like a lot of work! May their path be free of bugs and other critters.

**Comment by meisbellajack** [comment]
I completely understand the frustration regarding the branding shift. As someone deep in the Shopify and e-commerce ecosystem, the Remix → React Router v7 move felt like a hurdle, but the technical payoff in 2026 has been significant.

The core Remix philosophy loaders, actions, and progressive enhancement is now the foundation of React Router Framework Mode. This consolidation actually helped stabilize the tools we use to build high-performance apps.

For example, when we developed [Misk Smart Product Options](https://apps.shopify.com/misk-smart-product-options), we had to choose between staying in the 'legacy' script-injection world or embracing this new Native-First architecture. By following the React Router v7/Remix patterns, we were able to build a Native App Embed that:

Eliminates CLS: No more layout shifts that plague older Shopify apps.

Maintains Speed: It leverages the same parallel data loading that made Remix famous.

Future-Proof: Whether you call it Remix or React Router, the Web Standards approach ensures your store doesn't break when the next framework nap happens.

Ultimately, the name changed, but the performance-first mission stayed the same. It’s that mission that allows us to offer pro-level tools for $0 to early-stage founders today.

**Comment by aaronbutler1515-cyber** [comment]
I get the point you're making, and I agree that the underlying architecture
is much stronger now than the branding transition initially suggested.
The rename from Remix to React Router Framework Mode definitely created
confusion, especially for teams that had already invested heavily in the
Remix mental model and ecosystem. But in practice, most of the important
ideas — loaders, actions, nested routes, progressive enhancement, and
parallel data loading — survived intact and are arguably better integrated
now.
